### PR TITLE
Fast pump history

### PR DESCRIPTION
**Pull request by scottleibrand**
Started with #932 and made some tweaks to resolve merge conflicts, and improve logging and error handling. @alimhassam can you test this out?

### CODE REVIEW

File: bin/oref0-pump-loop.sh [hunk 1]
```diff
@@ -883,16 +846,51 @@ function check_tempbasal() {
...
+  fi
 }
-function read_pumphistory_24h() {
+function read_full_pumphistory() {
   set -o pipefail
-  pumphistory -n 27 2>&3 | jq -f openaps.jq | tee settings/pumphistory-24h-zoned.json 2>&3 >&4 \
-    && cat settings/pumphistory-24h-zoned.json | jq .[0].timestamp
+  rm monitor/pumphistory-24h-zoned.json
+  echo -n "Full history refresh" \
+  && ((( pumphistory -n 27 2>&3 | jq -f openaps.jq 2>&3 | tee monitor/pumphistory-24h-zoned.json 2>&3 >&4 ) \
```
Comment (alimhassam) [inline bin/oref0-pump-loop.sh]: Maybe this can be structured the same way as `read_fullpumphistory` using tryfail?

Comment (alimhassam) [reply bin/oref0-pump-loop.sh]: Actually there might be a small issue with this code. This just had happen to me luckily while i was semi-monitoring the logs and the pumphistory file size. 

My rig was on another floor with poor reception, and the daily full history refresh failed leaving the pumphistory file empty. It corrected on the next loop, but i'm wondering if we should save the result on a sepearete file and only overwrite on sucess?

Comment (scottleibrand) [reply bin/oref0-pump-loop.sh]: Yeah, that would probably be slightly better. Should be safe either way: as long as the pumphistory is left either old or empty it won’t loop, but for reporting leaving the old file unchanged is better.

**Comment by scottleibrand** [comment]
@alimhassam any idea how this is possible, or what we'd need to do to check for it?
```
Continuing oref0-pump-loop at Mon 26 Feb 22:50:41 PST 2018
Preflight OK. Profile less than 60m old; Profile valid. Retrying invoke_pumphistory_etc
Pump history updated through 2018-02-26T21:57:31-08:00; meal.json refreshed
Listening for 9 s silence: Listening: .No interfering pump comms detected from other rigs (this is a good thing!)
Continuing oref0-pump-loop at Mon 26 Feb 22:51:52 PST 2018
Checking pump clock: "2018-02-26T22:51:33-08:00" is within 90s of current time: Mon 26 Feb 22:51:53 PST 2018
```
There were lots of treatments from other rigs in the meantime.  It eventually did error out with:
```
{"iob":-0.248,"activity":-0.0006,"basaliob":-0.719,"bolusiob":0.471,"netbasalinsulin":-3.25,"bolusinsulin":8.2,"time":"2018-02-27T06:51:35.000Z","iobWithZeroTemp":{"iob":-0.248,"activity":-0.0006,"basaliob":-0.719,"bolusiob":0.471,"netbasalinsulin":-3.25,"bolusinsulin":8.2,"time":"2018-02-27T06:51:35.000Z"},"lastBolusTime":1519704402000,"lastTemp":{"rate":0.5,"timestamp":"2018-02-26T21:57:31-08:00","started_at":"2018-02-27T05:57:31.000Z","date":1519711051000,"duration":30}}
{"delta":1,"glucose":151,"noise":null,"short_avgdelta":1.28,"long_avgdelta":5.42,"date":1519714190000}
Autosens ratio: 0.75; Adjusting basal from 1.15 to 0.85; target_bg from 95 to 107; ISF from 39.7 to 53; CR: 10.528
currenttemp: { duration: 54, rate: 0, temp: 'absolute' } lastTempAge: 55 m tempModulus: 19 m
{"temp":"absolute","deliverAt":"2018-02-27T06:52:19.422Z","reason":"Warning: currenttemp rate 0 != lastTemp rate 0.5 from pumphistory; canceling temp","duration":0,"rate":0}
Pumphistory/temp mismatch: retrying
Couldn't smb_verify_suggested
oref0-pump-loop failed.
```
But it never should have gotten past the pumphistory refresh in the first place.

**Comment by alimhassam** [comment]
it seems like the errorcode from the processsubstitution didn't get propagated to the if statement. 
`<(pumphistory -s $topRecordTimestamp  2>&3 | jq -f openaps.jq 2>&3 ) `

I thought that `set -o pipefail` would do that, but just read it up on that now, it seems like it doesn't. We need to figure out how to get the errorcode propagated back to the if statement.

**Comment by efidoman** [comment]
I've been running on this branch all afternoon and it is working great. I have looked through the logs and have no errors with pumphistory updates. Nightscout is also working great (after I switched over to the new ns-loop).

**Comment by scottleibrand** [comment]
I think this is about ready to merge, so I'll do that soon if no objections, possibly this evening.

**Comment by alimhassam** [comment]
+1 happy to have this merge working for me

**Pull request merged**